### PR TITLE
Give meaningful error messages when checking response_t in ph5validate

### PR DESCRIPTION
**Pull request by damhuonglan**
In #334, Akram's 2nd and 3rd requests are to give meaningful error messages in ph5validate and ph5tostationxml for issue related to response table, so I created this branch that gives error logs for the following cases:

+ Index duplicated: "Response_t n_i(s) duplicated: 1,3,6"

+ Response n_i exist in array table but not in response table: "No response entry for n_i=6."

+ Wrong response_das_filename: "009-9001-1 response_table_n_i 4: Response das file name should be 'rt125a_200_1_32' instead of 'rt125a_500_1_32'."

+ Wrong response_sensor_filename: "008-8001-1 response_table_n_i 1: Response sensor file name should be 'cmg3t' instead of 'gs11v'."

+ Wrong response_das_filename while response_sensor_filename is empty: "002-0407-1 response_table_n_i 5: Response das file name should be 'NoneQ330_200_1_1' or 'NoneQ330_NoneCMG3T_200HHN' instead of 'NoneQ330_CMG3T_200HHN'."

+ No data loaded: "No response data loaded for cmg_3t."

### Checklist
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [x] All tests pass.
- [x] Any new or changed features have are documented.
- [x] Changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

### CODE REVIEW

**Comment by damhuonglan** [comment]
@dsentinel  Holly got error "list index out of range" when testing ph5_validate.  I figure out the bug is at these code lines
https://github.com/PIC-IRIS/PH5/blob/a99548f4329f49f3a06721172f38e63e9fbf73b4/ph5/utilities/ph5validate.py#L547-L549
I should change the code to:
```
  while check_end < check_start:
      try:
          i += 1
          check_end = das_time_list[index+i][0] - 1
      except IndexError:
          break
```

But because it isn't related to response table or any code changes in this PR, should I fix it here or create a new PR to fix it.

**Comment by damhuonglan** [comment]
@dsentinel: Please ignore the above comment. There are some errors I found when I check on ph5_validation.log that I think I have to create issue and open another branch to fix it. I will relate that issue to this PR when I create it.

**Comment by dsentinel** [comment]
Looks good to me pending:
@ascire-pic , @hrotman-pic, can you please review/confirm expected messages?
@damhuonglan , recently merged PR's have created a conflict here, please resolve.

**Comment by hrotman-pic** [comment]
Summary of what I have done and not done in reviewing this PR.


>     * Index duplicated: "Response_t n_i(s) duplicated: 1,3,6"

This is good.  This message outputs to the screen, and, for ph5_validate, is included in the log file.
> 
>     * Response n_i exist in array table but not in response table: "No response entry for n_i=6."

This is also good.  This message outputs to the screen and is included in the ph5_validate log file.
> 
>     * Wrong response_das_filename: "009-9001-1 response_table_n_i 4: Response das file name should be 'rt125a_200_1_32' instead of 'rt125a_500_1_32'."
> 
>     * Wrong response_sensor_filename: "008-8001-1 response_table_n_i 1: Response sensor file name should be 'cmg3t' instead of 'gs11v'."

I did not have an archive I was familiar with that had incorrect sensor or das filename in the array or response table, so I have not tested these two items.
> 
>     * Wrong response_das_filename while response_sensor_filename is empty: "002-0407-1 response_table_n_i 5: Response das file name should be 'NoneQ330_200_1_1' or 'NoneQ330_NoneCMG3T_200HHN' instead of 'NoneQ330_CMG3T_200HHN'."

I encountered this error while testing an archive that had response data loaded for some stations but not others.  In that test archive, all data was from Fairfield nodes, and the first row of the response table was correctly populated while the second row of the response table did not have any response information loaded.  The existing data was 500 sps, 36 dB, and the new data was 2000 sps, 24 dB.  The error message from ph5_validate was:
ERROR: 091-95001-1 response_table_n_i 0: Response das file name should be 'ZLAND3C_2000_1_36' or 'ZLAND3C__2000GP1' instead of 'ZLAND3C_500_1_36'.

For a response table that has one or more rows without any response data but does have response data in one or more other rows of the response table, I would expect an error that is something like:
"ERROR: ZLAND3C_2000_1_24 has no response loaded" or "ERROR: No response data loaded in response_table_n_i=1"

> 
>     * No data loaded: "No response data loaded for cmg_3t."

For the test archive I used that had no response data loaded, this case results in the error message: 
"ERROR: Response_t n_i=0: response_file_das_a is '' while das model exists." for a test archive with node data.  This tells the user that something is wrong.  I think that an error message similar to the example, "No response data loaded for cmg_3t" would provide additional help to users.
I have not tested a case with no response loaded for non-node data or for more than one instrument type.

Additionally, I ran ph5_validate on some PH5 volumes that have already been sent in to the DMC to verify that the changes did not break an already-validated archive.  I ran into no major issues with those archives.

**Comment by damhuonglan** [comment]
> I encountered this error while testing an archive that had response data loaded for some stations but not others. In that test archive, all data was from Fairfield nodes, and the first row of the response table was correctly populated while the second row of the response table did not have any response information loaded. The existing data was 500 sps, 36 dB, and the new data was 2000 sps, 24 dB. The error message from ph5_validate was:
> ERROR: 091-95001-1 response_table_n_i 0: Response das file name should be 'ZLAND3C_2000_1_36' or 'ZLAND3C__2000GP1' instead of 'ZLAND3C_500_1_36'.
This error message means that at the array 091- station 95001- channel 1, response_table_n_i is 0, while in response table at n_i=0 the field response_file_das_a is 'ZLAND3C_500_1_36'

> 
> For the test archive I used that had no response data loaded, this case results in the error message:
> "ERROR: Response_t n_i=0: response_file_das_a is '' while das model exists." for a test archive with node data. This tells the user that something is wrong. I think that an error message similar to the example, "No response data loaded for cmg_3t" would provide additional help to users.
This error message means that response_table_n_i from the array table is 0, and the response_file_das_a is '' for n_i=0.

All the errors are trace from array table to response table and report the status of the file name then the response data loaded. I can add additional check on the data loaded for that. It is simple for checking response_file_sensor_a but a little more difficult for response_file_das_a because the app cannot distinguish which one is created from metadatatoph5, which one is created from resp_load.

**Comment by damhuonglan** [comment]
Thank you for your approval on PR 383. I have a slightly change that hopefully make it clearer to user:

- To check if resp_load has been run for a ph5, ph5_validate/ph5tostationxml will check if field response_file_das_a has been filled for any response entry instead of check if all response_table_n_i in array table are 0 which makes user confused. If there is no response_file_das_a filled like in /external/ph5/Working/Holly_append_test/RAW/PH5_test/Sigma , ph5tostationxml will stop checking if resp_load hasn't been run and give the following error:
```
ERROR: All response file names are blank in response table. Check if resp_load has been run.
INFO: Request resulted in no data.
```
     In ph5_validate.log:

```
Response_t
1 error, 0 warning, 0 info
-=-=-=-=-=-=-=-=-
ERROR: All response file names are blank in response table. Check if resp_load has been run.
```

- In case there are some response_file_das_a filled but there are also some blank response_file_das/sensor_a while they are supposed to be filled, they will be reported like this
`ERROR: 001-1131-3 response_table_n_i 0: response_file_das_a is blank while das model exists.`

- Besides, before ph5_validate checking the response errors, the following instruction will be provided:
`INFO: NOTICE: The procedure of error checking for response table is getting the models and response_table_n_i from station entry in array table, from that response_table_n_i look for the corresponding n_i in response table and check if the response_file_das/response_a match with either metadatatoph5 format (dasModel_sensorModel_sampleRate_channelCode) or resp_load format (dasModel_sampleRate_sampleRateMultiplier_gain); then, last, check if the response data have been added for those response file names.`
I don't give this instruction in ph5tostationxml because DMC uses this too.

Please give me your opinion on the checking for resp_load. Thank you.

**Review by hrotman-pic — APPROVED** [review]
This looks good.  Error messages print to the log for validate, and print to stdout for validate and ph5tostationxml.  This includes the error messages added after my first review.

I could not reproduce the expected error message for some response file das_a filled but some blank, but there is an error message to alert the user.  I think it is better to keep this case in mind when working on an eventual successor to PH5.

Approving PR.  Thank you.

File: ph5/utilities/tests/test_ph5validate.py [hunk 1]
```diff
@@ -9,14 +9,91 @@
...
+
+    def tearDown(self):
+        self.ph5API_object.close()
+        super(TestPH5Validate_response, self).tearDown()
+
+    def test_check_array_t(self):
+        with LogCapture() as log:
+            log.setLevel(logging.ERROR)
+            vb_array, resp_check_info = self.ph5validate.check_array_t()
+            self.assertEqual(resp_check_info, self.resp_check_info)
```
Comment (MaevaP) [inline ph5/utilities/tests/test_ph5validate.py:63]: You don't seem to be using the log outputs in this specific unit test.

File: ph5/utilities/tests/test_validation.py [hunk 2]
```diff
@@ -1,12 +1,319 @@
...
+        self.ph5API_object.read_response_t()
+        response_t = self.ph5API_object.get_response_t_by_n_i(1)
+        response_t['n_i'] = 2
+        errors = set()
+        with LogCapture() as log:
+            log.setLevel(logging.ERROR)
+            validation.check_resp_unique_n_i(
+                self.ph5API_object, errors)
+        self.assertEqual(errors,
+                         {('Response_t n_i(s) duplicated: 2', 'error')})
```
Comment (MaevaP) [inline ph5/utilities/tests/test_validation.py:289]: Same here, you don't seem to use the log outputs in that particular test.

**Review by MaevaP — APPROVED** [review]
Looks good to me besides the two comments I have regarding the unused variable `log` in some of your unit tests.

**Pull request merged**